### PR TITLE
Tao 862 test taker review screen

### PR DESCRIPTION
**Pull request by jsconan**
Related to: https://oat-sa.atlassian.net/browse/TAO-862

An update is needed to be able to use the new test taker review screen.
But, you will need to manually enable the option in the file `config/taoQtiTest/testRunner.conf.php`.

There are some interesting options you can change:
- **progress-indicator** : allows to change the behavior of the progress bar (can be `percentage`or `position`
- **test-taker-review** : enables/disables the test taker review screen
- **test-taker-review-region** : set the region on which display the review screen (can be `left`or `right`)
- **test-taker-review-section-only** : when set to `true`, the review screen only display the current section
- **test-taker-review-prevents-unseen** : when set to `true`, only the viewed item can be accessed, the unseen items are simply disabled

On the review screen you can freely navigate between parts, sections and items, but only in non linear test.

To mark an item for review, just click on the icon before its name in the review screen.

### CODE REVIEW

File: actions/class.TestRunner.php [hunk 1]
```diff
@@ -289,6 +294,61 @@ public function index() {
...
+     *
+     */
+    public function jumpTo() {
+        $this->beforeAction();
+        $session = $this->getTestSession();
+
+        try {
+            $session->jumpTo(intval($this->getRequestParameter('position')));
+
+            if ($session->isRunning() === true && taoQtiTest_helpers_TestRunnerUtils::isTimeout($session) === false) {
```
Comment (krampstudio) [inline actions/class.TestRunner.php]: is triple = comparison required ?
JS habits ;)

File: views/js/controller/runtime/testRunner.js [hunk 2]
```diff
@@ -98,6 +99,36 @@ define([
                 iframeNotifier.parent('unloading');
             },
 
+            jump: function(position) {
```
Comment (krampstudio) [inline views/js/controller/runtime/testRunner.js]: jsDoc is missing

File: views/js/controller/runtime/testRunner.js [hunk 3]
```diff
@@ -98,6 +99,36 @@ define([
...
+                this.disableGui();
+                var that = this;
```
Comment (krampstudio) [inline views/js/controller/runtime/testRunner.js]: Our convention is to use `self`

File: views/js/controller/runtime/testRunner.js [hunk 4]
```diff
@@ -98,6 +99,36 @@ define([
                 iframeNotifier.parent('unloading');
             },
 
+            jump: function(position) {
+                this.disableGui();
```
Comment (krampstudio) [inline views/js/controller/runtime/testRunner.js]: variable to be declared first (we want hoisting to be reflected)

File: actions/class.TestRunner.php (hunk 1 above)
Comment (jsconan) [inline actions/class.TestRunner.php]: This part of code was copied from other navigation methods, and I assumed this kind of comparison was needed in that case

File: views/js/runner/testReview.js [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+
+            // build the component structure and inject it into the DOM
+            this.$container = $(element);
+            insertMethod = this.$container[insertMethod];
+            if (insertMethod) {
+                insertMethod.call(this.$container, navigatorTpl({
+                    region: putOnRight ? 'right' : 'left'
+                }));
+            } else {
+                throw "Unable to inject the component structure into the DOM";
```
Comment (krampstudio) [inline views/js/runner/testReview.js]: Please throw using `new Error('..')`

File: views/js/runner/testReview.js [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+            // access to the panel displayed when a linear part is reached
+            this.$linearState = this.$component.find('#qti-navigator-linear');
+        },
+
+        /**
+         * Installs the event handlers on the underlying DOM elements
+         * @private
+         */
+        _initEvents: function() {
+            var that = this;
```
Comment (krampstudio) [inline views/js/runner/testReview.js]: we use `self` by convetion

File: views/js/runner/testReview.js [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+         * Initializes the component
+         * @param {String|jQuery|HTMLElement} element The element on which install the component
+         * @param {Object} [options] A list of extra options
+         * @param {String} [options.region] The region on which put the component: left or right
+         * @param {Boolean} [options.sectionOnly] Limit the review screen to the current test section only
+         * @param {Boolean} [options.preventsUnseen] Prevents the test taker to access unseen items
+         * @returns {testReview}
+         */
+        init: function init(element, options) {
+            this.options = _.isObject(options) && options || {};
```
Comment (krampstudio) [inline views/js/runner/testReview.js]: relfeclt hoisting in variable delcaration

File: views/templates/test_runner.tpl [hunk 8]
```diff
@@ -1,6 +1,14 @@
...
 use oat\tao\helpers\Template;
 use oat\tao\helpers\Layout;
+
+$contentPanelSetup = '';
+if (get_data('review_screen')) {
+    $contentPanelSetup .= ' data-review-screen="1"';
+}
+if (get_data('review_region')) {
+    $contentPanelSetup .= ' data-review-region="' . get_data('review_region') . '"';
+}
```
Comment (krampstudio) [inline views/templates/test_runner.tpl]: There shouldn't be any construction in the template expect var,  if/then/else and for loops,

**Comment by krampstudio** [comment]
I've an issue with all the ids in the template. 
I suggest you to use only one id (your top container or none if the container is given to factory) and then uses only the classes to access the elements.

**Comment by krampstudio** [comment]
Last but not least, the testReview should be tested.

File: views/scss/navigator.scss [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        }
+    }
+
+    .qti-navigator-counter {
+        text-align: right;
+        float: right;
+        @include font-size(12);
+    }
+
+    .qti-navigator-message {
```
Comment (krampstudio) [inline views/scss/navigator.scss]: empty selector

**Comment by krampstudio** [comment]
Non blocking comments :
- use a media query to resize the the navigation screen
- in low resolution, modal feedback appears behind the navigation screen (z-index issue)
- message when trying to exit can be confusing
- the current item in the navigation should be highlighted (here the gray seems to be not enough)

**Pull request merged**